### PR TITLE
Use a custom deserializer instead of wrapper types

### PR DESCRIPTION
**Pull request by romac**

### CODE REVIEW

**Comment by rnbguy** [comment]
To document how I solved the `num_bigin::BigInt` deserialization.

The ITF trace deserialization to native types works as follows,
1. deserialize raw `itf` JSON to `serde_json::Value`
2. then deserializing `serde_json::Value` to `itf::value::Value` (`impl<'de> Deserialize<'de> for itf::value::Value`) 
3. then deserializing `itf::value::Value` to `T` (`impl<'de> Deserializer<'de> for itf::value::Value`)

To get native deserialization of `num_bigint::BigInt`, `itf::value::Value::BigInt` needs to be deserialized as `(int, [uint])`. I added https://github.com/informalsystems/itf-rs/pull/6/commits/a56363c5e86d1cf30bd3c6aa651914cc289ad13b for this. And, https://github.com/informalsystems/itf-rs/pull/6/commits/92bb9d3d12adf7e4b26b0119385d8f5b2415fcbc allows `itf::value::Value::BigInt` to be deserialized as a tuple.

But this creates an weird issue - `itf::from_str::<itf::value::Value>` will deserialize `{"#bigint": "99"}` as `[1, [99]]`. As `itf::from_str` will deserialize itf-bigint format to `itf::value::Value` and then again to itself. The second deserialization makes it a tuple and `itf::value::Value` does not have any information on how to deserialize it back to `itf::value::Value::BigInt`. So I added https://github.com/informalsystems/itf-rs/pull/6/commits/c03bec1122fbb533ffe19e6c5dcac1c3465cfb82.

Note, with this, `itf::from_str` does not differentiate between `{"#bigint": "99"}` and `[1, [99]]`; they both can be deserialized to `num_bigint::BigInt`.

**Comment by rnbguy** [comment]
Note, this does not support `num_bigint::BigUint` which has a different deserialization format than `num_bigint::BigInt`.

- `BigInt`: `(int, [uint])`
- `BigUint`: `[uint]`

**Comment by rnbguy** [comment]
`Value::deserialize(Value)` is creating equivalency issues for `Value::{Set, Map, List}` too. I added testcases in https://github.com/informalsystems/itf-rs/pull/6/commits/d89d2fb5133b61e18d33a62971168e8ecd0fa797.

I suggest making `itf::Value` private. This way we can also revert https://github.com/informalsystems/itf-rs/commit/c03bec1122fbb533ffe19e6c5dcac1c3465cfb82 :slightly_smiling_face:

**Comment by romac** [comment]
Amazing stuff @rnbguy! Thanks so much!

> I suggest making itf::Value private. This way we can also revert https://github.com/informalsystems/itf-rs/commit/c03bec1122fbb533ffe19e6c5dcac1c3465cfb82 🙂

Yeah let's do that!

File: itf/tests/regression.rs [hunk 1]
```diff
@@ -124,3 +124,42 @@ fn test_map_with_non_str_key() {
     let value = serde_json::from_value::<itf::Value>(itf).unwrap();
     itf::Value::deserialize(value).unwrap();
 }
+
+#[test]
+fn test_complete() {
```
Comment (romac) [inline itf/tests/regression.rs:140]: Nice!

File: itf/src/de.rs [hunk 2]
```diff
@@ -258,6 +258,7 @@ impl<'de> Deserializer<'de> for Value {
     {
         match self {
             Value::BigInt(v) => visit_bigint(v, visitor),
+            Value::Number(v) => visit_bigint(BigInt::new(v), visitor),
```
Comment (romac) [inline itf/src/de.rs:261]: Good idea!

**Comment by rnbguy** [comment]
@romac any idea why this (https://github.com/informalsystems/itf-rs/pull/6/commits/9920493b3a74865f1d41aca72838a6e57d48d9b5) is not working?

https://github.com/informalsystems/itf-rs/blob/9920493b3a74865f1d41aca72838a6e57d48d9b5/itf/tests/regression.rs#L141-L154

**Comment by romac** [comment]
Ah good catch! Yeah it's a shame we can't seem to support this.

That said, we can perhaps still enable this by providing a `from_bigint<'de, A: TryFrom<BigInt>, D>(d: D) -> Result<A, D::Error>` helper function which could be used with `#[serde(deserialize_with = "itf::de::from_bigint")]`.

See https://github.com/informalsystems/itf-rs/pull/6/commits/a2eef53c28f230d1ec0ec24578815f7b2ab3b132

**Comment by rnbguy** [comment]
Yea `serde(with)` is the last straw :sweat_smile: 

Btw. I would prefer `serde_with::DeserializeAs` with `serde_with::As`. This adds a dependency to `serde_with`. But this way, we can support container types too.

```rs
use itf::value::BigInt;
use serde_with::As;

impl<'de, T> serde_with::DeserializeAs<'de, T> for BigInt { ... }

#[derive(Deserialize, Debug)]
#[serde(tag = "typ")]
enum FooBarWithInt {
    Foo {
        #[serde(with = "As::<Vec<Vec<BigInt>>>")]
        _foo: Vec<Vec<i64>>,
    },
    Bar {
        _bar: String,
    },
}
```

I can take care of the changes tomorrow :slightly_smiling_face:

**Comment by romac** [comment]
Ah good point, didn't think of containers! I am fine with adding `serde-with` as a dep :)

**Comment by rnbguy** [comment]
I removed the code that implicitly serialized `BigInt` to `(u)int` and vice-versa. Now, `{"#bigint": "<num>"}` is always deserialized to `BigInt`.

For rust primitive integer types, a user has to pass `serde_with::As` type in _[serde-with](https://serde.rs/field-attrs.html#with)_ (e.g. `#[serde(with = "As::<itf::de::Integer>")]`) or `serde_with::serde_as` attribute accordingly.

**Comment by rnbguy** [comment]
Btw, `serde_with` provides `TryFromInto` - which can be used as follows - `#[serde(with = "As::<TryFromInt<BigInt>>")]` in our context. `itf::de::Integer` is just a shorthand for it.

**Comment by romac** [comment]
Awesome!

**Comment by rnbguy** [comment]
Btw, maybe it make sense to redefine `Integer` as just

```rs
pub type Integer = TryFromInto<BigInt>;
```

**Pull request merged**